File: backend/domain/catalogo_correcciones.py
"""
Catálogo de campos identificables en una solicitud de devolución.

Fuente única de verdad sobre qué campos del formulario pueden ser referenciados
explícitamente por un área al devolver un expediente para corrección.

Agregar nuevos campos aquí es la única acción necesaria para extender el sistema.
"""

from __future__ import annotations

# IDs de campo válidos que un área puede marcar como "requiere corrección".
# Deben coincidir exactamente con los IDs usados en el catálogo del frontend
# (frontend/src/data/catalogoCorrecciones.js).
CAMPOS_IDENTIFICABLES_PARA_CORRECION: frozenset[str] = frozenset({
    # Paso 2 — Información básica
    "tipo_contraparte",
    "tipo_persona",
    "tipo_solicitud",
    "clasificacion_actividad",
    "razon_social",
    "tipo_identificacion",
    "numero_identificacion",
    "digito_verificacion",
    "direccion",
    "pais",
    "departamento",
    "ciudad",
    "telefono",
    "fax",
    "correo",
    "codigo_ica",
    "pagina_web",

    # Paso 3 — Representante legal
    "nombre_representante",
    "tipo_doc_representante",
    "numero_doc_representante",
    "fecha_expedicion",
    "ciudad_expedicion",
    "nacionalidad",
    "fecha_nacimiento",
    "ciudad_nacimiento",
    "profesion",
    "correo_representante",
    "telefono_representante",
    "direccion_funciones",
    "pais_funciones",
    "departamento_funciones",
    "ciudad_funciones",
    "ciudad_residencia",
    "direccion_residencia",

    # Paso 4 — Junta, accionistas, beneficiarios (secciones de tabla)
    "junta_directiva",
    "accionistas",
    "beneficiario_final",

    # Paso 5 — Información financiera
    "actividad_economica",
    "codigo_ciiu",
    "ingresos_mensuales",
    "otros_ingresos",
    "egresos_mensuales",
    "total_activos",
    "total_pasivos",
    "patrimonio",

    # Paso 6 — Operaciones en moneda extranjera
    "realiza_operaciones_moneda_extranjera",
    "paises_operaciones",
    "tipos_transaccion",
    "referencias_comerciales",
    "referencias_bancarias",

    # Paso 7 — Clasificación y contactos
    "actividad_clasificacion",
    "sector",
    "superintendencia",
    "contacto_ordenes_nombre",
    "contacto_ordenes_cargo",
    "contacto_ordenes_telefono",
    "contacto_ordenes_correo",
    "contacto_pagos_nombre",
    "contacto_pagos_cargo",
    "contacto_pagos_telefono",
    "contacto_pagos_correo",
    "informacion_bancaria_pagos",

    # Paso 8 — Declaraciones y firma
    "origen_fondos",
    "dia_firma",
    "mes_firma",
    "year_firma",
    "ciudad_firma",
})


# Etiquetas legibles de cada campo identificable.
# Se usan en el correo de notificación para que la contraparte entienda
# qué campos corregir sin necesidad de abrir el formulario.
ETIQUETA_DE_CAMPO_EN_CORRECION: dict[str, str] = {
    "tipo_contraparte":        "Tipo de Contraparte",
    "tipo_persona":            "Tipo de Persona",
    "tipo_solicitud":          "Tipo de Solicitud",
    "clasificacion_actividad": "Clasificación de Actividad",
    "razon_social":            "Razón Social / Nombre",
    "tipo_identificacion":     "Tipo de Identificación",
    "numero_identificacion":   "Número de Identificación / NIT",
    "digito_verificacion":     "Dígito de Verificación",
    "direccion":               "Dirección",
    "pais":                    "País",
    "departamento":            "Departamento",
    "ciudad":                  "Ciudad",
    "telefono":                "Teléfono",
    "fax":                     "Fax",
    "correo":                  "Correo Electrónico",
    "codigo_ica":              "Código ICA",
    "pagina_web":              "Página Web",
    "nombre_representante":    "Nombre del Representante Legal",
    "tipo_doc_representante":  "Tipo de Documento del Representante",
    "numero_doc_representante":"Número de Documento del Representante",
    "fecha_expedicion":        "Fecha de Expedición del Documento",
    "ciudad_expedicion":       "Ciudad de Expedición",
    "nacionalidad":            "Nacionalidad",
    "fecha_nacimiento":        "Fecha de Nacimiento",
    "ciudad_nacimiento":       "Ciudad de Nacimiento",
    "profesion":               "Profesión",
    "correo_representante":    "Correo del Representante",
    "telefono_representante":  "Teléfono del Representante",
    "direccion_funciones":     "Dirección donde desempeña funciones",
    "pais_funciones":          "País de funciones",
    "departamento_funciones":  "Departamento de funciones",
    "ciudad_funciones":        "Ciudad de funciones",
    "ciudad_residencia":       "Ciudad de Residencia (Persona Natural)",
    "direccion_residencia":    "Dirección de Residencia (Persona Natural)",
    "junta_directiva":         "Miembros de Junta Directiva",
    "accionistas":             "Accionistas",
    "beneficiario_final":      "Beneficiarios Finales",
    "actividad_economica":     "Actividad Económica",
    "codigo_ciiu":             "Código CIIU",
    "ingresos_mensuales":      "Ingresos Mensuales",
    "otros_ingresos":          "Otros Ingresos",
    "egresos_mensuales":       "Egresos Mensuales",
    "total_activos":           "Total Activos",
    "total_pasivos":           "Total Pasivos",
    "patrimonio":              "Patrimonio",
    "realiza_operaciones_moneda_extranjera": "¿Realiza Operaciones en Moneda Extranjera?",
    "paises_operaciones":      "Países de Operaciones",
    "tipos_transaccion":       "Tipos de Transacción",
    "referencias_comerciales": "Referencias Comerciales",
    "referencias_bancarias":   "Referencias Bancarias",
    "actividad_clasificacion": "Actividad / Clasificación",
    "sector":                  "Sector Económico",
    "superintendencia":        "Superintendencia",
    "contacto_ordenes_nombre": "Contacto Órdenes — Nombre",
    "contacto_ordenes_cargo":  "Contacto Órdenes — Cargo",
    "contacto_ordenes_telefono": "Contacto Órdenes — Teléfono",
    "contacto_ordenes_correo": "Contacto Órdenes — Correo",
    "contacto_pagos_nombre":   "Contacto Pagos — Nombre",
    "contacto_pagos_cargo":    "Contacto Pagos — Cargo",
    "contacto_pagos_telefono": "Contacto Pagos — Teléfono",
    "contacto_pagos_correo":   "Contacto Pagos — Correo",
    "informacion_bancaria_pagos": "Información Bancaria de Pagos",
    "origen_fondos":           "Origen de Fondos",
    "dia_firma":               "Día de Firma",
    "mes_firma":               "Mes de Firma",
    "year_firma":              "Año de Firma",
    "ciudad_firma":            "Ciudad de Firma",
}


def resolver_etiquetas(campos: list[str]) -> list[str]:
    """Convierte IDs de campo a sus etiquetas legibles para mostrar en el correo."""
    return [ETIQUETA_DE_CAMPO_EN_CORRECION.get(campo, campo) for campo in campos]


def validar_campos_identificados(campos: list[str]) -> list[str]:
    """
    Verifica que todos los IDs de campo proporcionados existen en el catálogo.

    Raises:
        ValueError: si algún ID no es reconocido, listando los inválidos.
    """
    invalidos = [c for c in campos if c not in CAMPOS_IDENTIFICABLES_PARA_CORRECION]
    if invalidos:
        raise ValueError(
            f"Los siguientes campos no son identificables en el formulario: {invalidos}. "
            f"Consulte el catálogo de campos permitidos."
        )
    return campos

File: frontend/src/data/catalogoCorrecciones.js
/**
 * Catálogo de campos identificables en una solicitud de devolución.
 *
 * Espeja backend/domain/catalogo_correcciones.py — los IDs deben ser idénticos.
 * Agregar nuevos campos aquí (y en el backend) es la única acción necesaria
 * para extender el sistema a nuevos campos sin tocar componentes.
 *
 * Estructura:
 *   paso       — número de paso del formulario multipágina
 *   etiqueta   — nombre de la sección que ve el usuario del portal
 *   campos     — lista de { id, etiqueta } ordenada como aparece en el formulario
 */

export const CATALOGO_CORRECCIONES = [
  {
    paso: 2,
    etiqueta: 'Información básica',
    campos: [
      { id: 'tipo_contraparte',       etiqueta: 'Tipo de Contraparte' },
      { id: 'tipo_persona',           etiqueta: 'Tipo de Persona' },
      { id: 'tipo_solicitud',         etiqueta: 'Tipo de Solicitud' },
      { id: 'clasificacion_actividad',etiqueta: 'Clasificación de Actividad' },
      { id: 'razon_social',           etiqueta: 'Razón Social / Nombre' },
      { id: 'tipo_identificacion',    etiqueta: 'Tipo de Identificación' },
      { id: 'numero_identificacion',  etiqueta: 'Número de Identificación / NIT' },
      { id: 'digito_verificacion',    etiqueta: 'Dígito de Verificación' },
      { id: 'direccion',              etiqueta: 'Dirección' },
      { id: 'pais',                   etiqueta: 'País' },
      { id: 'departamento',           etiqueta: 'Departamento' },
      { id: 'ciudad',                 etiqueta: 'Ciudad' },
      { id: 'telefono',               etiqueta: 'Teléfono' },
      { id: 'fax',                    etiqueta: 'Fax' },
      { id: 'correo',                 etiqueta: 'Correo Electrónico' },
      { id: 'codigo_ica',             etiqueta: 'Código ICA' },
      { id: 'pagina_web',             etiqueta: 'Página Web' },
    ],
  },
  {
    paso: 3,
    etiqueta: 'Representante legal',
    campos: [
      { id: 'nombre_representante',       etiqueta: 'Nombre Completo' },
      { id: 'tipo_doc_representante',     etiqueta: 'Tipo de Documento' },
      { id: 'numero_doc_representante',   etiqueta: 'Número de Documento' },
      { id: 'fecha_expedicion',           etiqueta: 'Fecha de Expedición' },
      { id: 'ciudad_expedicion',          etiqueta: 'Ciudad de Expedición' },
      { id: 'nacionalidad',               etiqueta: 'Nacionalidad' },
      { id: 'fecha_nacimiento',           etiqueta: 'Fecha de Nacimiento' },
      { id: 'ciudad_nacimiento',          etiqueta: 'Ciudad de Nacimiento' },
      { id: 'profesion',                  etiqueta: 'Profesión' },
      { id: 'correo_representante',       etiqueta: 'Correo Electrónico' },
      { id: 'telefono_representante',     etiqueta: 'Teléfono' },
      { id: 'direccion_funciones',        etiqueta: 'Dirección donde desempeña funciones' },
      { id: 'pais_funciones',             etiqueta: 'País de funciones' },
      { id: 'departamento_funciones',     etiqueta: 'Departamento de funciones' },
      { id: 'ciudad_funciones',           etiqueta: 'Ciudad de funciones' },
      { id: 'ciudad_residencia',          etiqueta: 'Ciudad de Residencia (Persona Natural)' },
      { id: 'direccion_residencia',       etiqueta: 'Dirección de Residencia (Persona Natural)' },
    ],
  },
  {
    paso: 4,
    etiqueta: 'Junta, accionistas y beneficiarios',
    campos: [
      { id: 'junta_directiva',   etiqueta: 'Miembros de Junta Directiva' },
      { id: 'accionistas',       etiqueta: 'Accionistas' },
      { id: 'beneficiario_final',etiqueta: 'Beneficiarios Finales' },
    ],
  },
  {
    paso: 5,
    etiqueta: 'Información financiera',
    campos: [
      { id: 'actividad_economica',  etiqueta: 'Actividad Económica' },
      { id: 'codigo_ciiu',          etiqueta: 'Código CIIU' },
      { id: 'ingresos_mensuales',   etiqueta: 'Ingresos Mensuales' },
      { id: 'otros_ingresos',       etiqueta: 'Otros Ingresos' },
      { id: 'egresos_mensuales',    etiqueta: 'Egresos Mensuales' },
      { id: 'total_activos',        etiqueta: 'Total Activos' },
      { id: 'total_pasivos',        etiqueta: 'Total Pasivos' },
      { id: 'patrimonio',           etiqueta: 'Patrimonio' },
    ],
  },
  {
    paso: 6,
    etiqueta: 'Operaciones y referencias',
    campos: [
      { id: 'realiza_operaciones_moneda_extranjera', etiqueta: '¿Realiza Operaciones en Moneda Extranjera?' },
      { id: 'paises_operaciones',    etiqueta: 'Países de Operaciones' },
      { id: 'tipos_transaccion',     etiqueta: 'Tipos de Transacción' },
      { id: 'referencias_comerciales', etiqueta: 'Referencias Comerciales' },
      { id: 'referencias_bancarias', etiqueta: 'Referencias Bancarias' },
    ],
  },
  {
    paso: 7,
    etiqueta: 'Clasificación y contactos',
    campos: [
      { id: 'actividad_clasificacion',      etiqueta: 'Actividad / Clasificación' },
      { id: 'sector',                       etiqueta: 'Sector Económico' },
      { id: 'superintendencia',             etiqueta: 'Superintendencia' },
      { id: 'contacto_ordenes_nombre',      etiqueta: 'Contacto Órdenes — Nombre' },
      { id: 'contacto_ordenes_cargo',       etiqueta: 'Contacto Órdenes — Cargo' },
      { id: 'contacto_ordenes_telefono',    etiqueta: 'Contacto Órdenes — Teléfono' },
      { id: 'contacto_ordenes_correo',      etiqueta: 'Contacto Órdenes — Correo' },
      { id: 'contacto_pagos_nombre',        etiqueta: 'Contacto Pagos — Nombre' },
      { id: 'contacto_pagos_cargo',         etiqueta: 'Contacto Pagos — Cargo' },
      { id: 'contacto_pagos_telefono',      etiqueta: 'Contacto Pagos — Teléfono' },
      { id: 'contacto_pagos_correo',        etiqueta: 'Contacto Pagos — Correo' },
      { id: 'informacion_bancaria_pagos',   etiqueta: 'Información Bancaria de Pagos' },
    ],
  },
  {
    paso: 8,
    etiqueta: 'Declaraciones y firma',
    campos: [
      { id: 'origen_fondos', etiqueta: 'Origen de Fondos' },
      { id: 'dia_firma',     etiqueta: 'Día de Firma' },
      { id: 'mes_firma',     etiqueta: 'Mes de Firma' },
      { id: 'year_firma',    etiqueta: 'Año de Firma' },
      { id: 'ciudad_firma',  etiqueta: 'Ciudad de Firma' },
    ],
  },
];

/**
 * Mapa plano id → etiqueta para consultas rápidas.
 * Construido una sola vez al cargar el módulo.
 */
export const ETIQUETA_POR_CAMPO = Object.fromEntries(
  CATALOGO_CORRECCIONES.flatMap(grupo => grupo.campos.map(campo => [campo.id, campo.etiqueta])),
);

/**
 * Mapa id → número de paso para navegación directa desde la corrección.
 */
export const PASO_POR_CAMPO = Object.fromEntries(
  CATALOGO_CORRECCIONES.flatMap(grupo => grupo.campos.map(campo => [campo.id, grupo.paso])),
);
